File: NumericalIntegral/simpson.jl
"""
    simpson(func::Function, a::Real, b::Real)

Numerically approximate the integral of func from `a` to `b` using Simpson's rule.

See also [`comsimp`], [`adapsimp`].

# Arguments

* `func::Function`: `func` is the integrand.
* `a::Real`: `a` is the lower limit of integration by default.
* `b::Real`: `b` is the supper limit of integration by defautlt.

If `a = b`, throw a Warning and return 0; if `a > b`, swap `a` and `b` to ensure `h = (b - a)/2 > 0`.

# Outputs

* `s::Float64`: s is the approximation of integral.

# Examples

```julia-repl
julia> simpson(cos, 0, 0)
Warning: The limits of integration are equal!
0

julia> simpson(sin, 0, pi/2)
1.0022798774922104

julia> simpson(x->exp(-x^2), 1, 0)
0.7471804289095104
```
"""
function simpson(func::Function, a::T, b::T) where T<:Real
    if a == b
        @warn "The limits of integration are equal!"
        return 0
    elseif a > b
        a, b = b, a
    end
    h = (b - a) / 2;
    s = h * (func(a) + 4 * func(a+h) + func(b)) / 3
    return s
end

File: NumericalIntegral/simpson.m
function s = simpson(func, a, b)
%SIMPSON - Simpson 公式数值近似求解函数 func 在区间 [a,b] 上的定积分.
%
% 语法: s = simpson(func, a, b)
%
% 输入:
%   func - 被积函数. func 是一个函数句柄或者匿名函数.
%   a - double. a 默认为积分下限.
%   b - double. b 默认为积分上限.
%       如果 a = b, 抛出积分上下限相等的警告并赋 s = 0.
%       如果 a > b, 交换 a 和 b 的值保证 h = (b - a) / 2 > 0.
%
% 输出:
%   s - double. s 是积分的近似值.
%
% 示例:
% > simpson(@cos, 0, 0)
% 警告! 积分上下限相等!
% ans =
%   0
% > simpson(@sin, 0, pi/2)
% ans =
%   1.002279877492210
% > simpson(@(x) exp(-x^2), 1, 0)
% ans =
%   0.747180428909510
%
% 另见: comsimp, adapsimp
%
%
    if a == b
        disp("警告! 积分上下限相等!");
        s = 0;
        return;
    elseif a > b
        c = a;
        a = b;
        b = c;
    end
    h = (b - a) / 2;
    s = h * (func(a) + 4 * func(a + h) + func(b)) / 3;
end
